### DISCUSSION TITLE
Type mismatch: inferred type is ContentView2 but () -> View was expected

### DISCUSSION THREAD

**Discussion by swift-jpg** [skiptools/skip#159] [Q&A] [answered]
Hi, I'm quite new to development with Swift and SwiftUI so apologies if this is a bad question.

I have been playing around with Skip recently, and I wanted to have two views: `ContentView` and `ContentView2`.

On the entry point of the program and in the public struct `RootView`, calling `ContentView` and `ContentView2` both work. However, in `ContentView`, using a `NavigationLink` to call `ContentView2()`, the preview of the app works fine, but when attempting to build the app it results in a compile error: `Type mismatch: inferred type is ContentView2 but () -> View was expected`.

Poking around files and a few google searches makes me feel that this is a Kotlin error and not a Swift problem.

My files:
ContentView.swift (Preview works)
```swift
import SwiftUI
public struct ContentView: View {
    public var body: some View {
        VStack(spacing: 16.0) {
            NavigationStack{
                NavigationLink(destination: ContentView2()){
                    Label("Click Me", systemImage: "hand.wave")}
                }
            }
            .padding()
        }
    }

        
#Preview {
    ContentView()
}
```

ContentView2.swift (Preview works)
```swift
import SwiftUI

public struct ContentView2: View {
    public var body: some View {
        Text("Hello, World!")
    }
}

#Preview {
    ContentView2()
}
```

HelloSkipApp (entry point file)

```swift
import Foundation
import OSLog
import SwiftUI

let logger: Logger = Logger(subsystem: "com.idk.HelloSkip", category: "HelloSkip")

/// The Android SDK number we are running against, or `nil` if not running on Android
let androidSDK = ProcessInfo.processInfo.environment["android.os.Build.VERSION.SDK_INT"].flatMap({ Int($0) })

/// The shared top-level view for the app, loaded from the platform-specific App delegates below.
///
/// The default implementation merely loads the `ContentView` for the app and logs a message.
public struct RootView : View {
    public init() {
    }

    public var body: some View {
        ContentView()
            .task {
                logger.log("Welcome to Skip on \(androidSDK != nil ? "Android" : "Darwin")!")
                logger.warning("Skip app logs are viewable in the Xcode console for iOS; Android logs can be viewed in Studio or using adb logcat")
            }
        
    }
}

#if !SKIP
public protocol HelloSkipApp : App {
}

/// The entry point to the HelloSkip app.
/// The concrete implementation is in the HelloSkipApp module.
public extension HelloSkipApp {
    var body: some Scene {
        WindowGroup {
            RootView()
        }
    }
}
```

Does this involve Creating a Multi-Module App in the Skip Documentation or adding Targets in Xcode? How can I get my build to succeed?

Thanks in advance and have a great day ahead! 

NOTE: I'm not sure if this is related, but after installing Skip when creating a new swift file in Xcode, I no longer have an option to rename the swift file before it is created. Could this be causing issue?

**Comment by marcprux** [comment, marked as answer]
Rather than doing this:

> ```swift
>                 NavigationLink(destination: ContentView2()){
>                     Label("Click Me", systemImage: "hand.wave")}
>                 }
> ```

I think you need to do this (note the extra braces):

> ```swift
>                 NavigationLink(destination: { ContentView2() }){
>                     Label("Click Me", systemImage: "hand.wave")}
>                 }
> ```

This is due to difficulty Skip has with supporting the autoclosure-style argument for the `NavigationLink`. Please try that out and let us know if it helps.

> NOTE: I'm not sure if this is related, but after installing Skip when creating a new swift file in Xcode, I no longer have an option to rename the swift file before it is created. Could this be causing issue?

I'm not sure what you mean by this. Do you mean Xcode's "Choose a template for your new file:" dialog that results from `File`/`New`/`File…` Xcode menu? I am able to rename the file when I use that workflow. Skip is unlikely to be affecting that aspect of Xcode.

**Reply by swift-jpg** [reply]
Adding the extra braces did help solve the issues and I could compile successfully. Thank you! One more thing: was this difficulty with supporting the autoclosure-style argument documented anywhere? I couldn't seem to find anything relating to this in the skip.tools documentation page.

> I'm not sure what you mean by this. Do you mean Xcode's "Choose a template for your new file:" dialog that results from File/New/File… Xcode menu? I am able to rename the file when I use that workflow. Skip is unlikely to be affecting that aspect of Xcode.

It's a funny minor issue, but it seems to only be affecting Skip projects.

**Basic Xcode project**

https://github.com/skiptools/skip/assets/173980614/8dbb988c-138f-48f9-af5e-c708386dd63b


**Skip project**


https://github.com/skiptools/skip/assets/173980614/3219ae0b-e172-4e65-a50a-366a9d6cdff9

**Reply by aabewhite** [reply]
The SwiftUI doc (i.e. if you open NavigationLink in Xcode) shows the non-closure destination constructor as deprecated (though without a version, so it's sort of a soft deprecation), and Skip generally doesn't support deprecated API:

```
    @available(iOS, introduced: 13.0, deprecated: 100000.0, message: "Pass a closure as the destination")
    @available(macOS, introduced: 10.15, deprecated: 100000.0, message: "Pass a closure as the destination")
    @available(tvOS, introduced: 13.0, deprecated: 100000.0, message: "Pass a closure as the destination")
    @available(watchOS, introduced: 6.0, deprecated: 100000.0, message: "Pass a closure as the destination")
    @available(visionOS, introduced: 1.0, deprecated: 100000.0, message: "Pass a closure as the destination")
    public init(destination: Destination, @ViewBuilder label: () -> Label)
```
